### PR TITLE
Feature: Add LocalTime Scalar

### PR DESCRIPTION
**Pull request by dmoree**
# Description

Adds a `LocalTime` scalar. It is passed in as a time string without timezone information and stored in the database as a neo4j `LocalTime` instance.

# Checklist

The following requirements should have been met (depending on the changes in the branch):

- [x] Documentation has been updated
- [x] TCK tests have been updated
- [x] Integration tests have been updated
- [ ] Example applications have been updated
- [x] New files have copyright header
- [x] CLA (https://neo4j.com/developer/cla/) has been signed

### CODE REVIEW

**Comment by neo4j-team-graphql** [comment]
Looks like you've updated the documentation!

Check out your changes at https://neo4j-graphql-docs-451.surge.sh

File: packages/graphql/src/schema/scalars/LocalTime.ts [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        throw new TypeError(`Value must be formatted as LocalTime: ${value}`);
+    }
+
+    return value;
+};
+
+export const parseLocalTime = (value: any) => {
+    const validatedValue = validateLocalTime(value);
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+    const { hour, minute, second, fraction } = LOCAL_TIME_REGEX.exec(validatedValue)!.groups!;
```
Comment (darrellwarde) [inline packages/graphql/src/schema/scalars/LocalTime.ts:40]: Same suggestion as on #449 - happy to discuss!

```suggestion
export const parseLocalTime = (value: any) => {
    if (typeof value !== "string") {
        throw new TypeError(`Value must be of type string: ${value}`);
    }
    
    const match = LOCAL_TIME_REGEX.exec(value);
    
    if (!match) {
        throw new TypeError(`Value must be formatted as LocalTime: ${value}`);
    }
    
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const { hour, minute, second, fraction } = match.groups!;
```

**Review by darrellwarde — CHANGES_REQUESTED** [review]
Thanks for also doing `LocalTime` - nearly there on all of the temporal types now apart from `Duration`! 😄  The same comment as on #449 left.

**Review by danstarns — CHANGES_REQUESTED** [review]
Hey @dmoree, Thanks for this. I would like to see something like `localTimeFields` [on the Node](https://github.com/neo4j/graphql/blob/master/packages/graphql/src/classes/Node.ts#L84) and Relationship classes. Then these fields can get picked up in particular places such as [specifying the field has auth on it](https://github.com/neo4j/graphql/blob/f16d2a27b4bc8be088e149658619190379b597d4/packages/graphql/src/classes/Node.ts#L143) and [allowing one to coalesce the field](https://github.com/neo4j/graphql/blob/master/packages/graphql/src/translate/where/create-relationship-where-and-params.ts#L59)(open question?).

**Comment by darrellwarde** [comment]
> Hey @dmoree, Thanks for this. I would like to see something like `localTimeFields` [on the Node](https://github.com/neo4j/graphql/blob/master/packages/graphql/src/classes/Node.ts#L84) and Relationship classes. Then these fields can get picked up in particular places such as [specifying the field has auth on it](https://github.com/neo4j/graphql/blob/f16d2a27b4bc8be088e149658619190379b597d4/packages/graphql/src/classes/Node.ts#L143) and [allowing one to coalesce the field](https://github.com/neo4j/graphql/blob/master/packages/graphql/src/translate/where/create-relationship-where-and-params.ts#L59)(open question?).

Could they not all be bundled into `dateTimeFields` (maybe rename it `temporalFields`)? I don't think there's really going to be any behaviour differences between the various types.

**Comment by danstarns** [comment]
> > Hey @dmoree, Thanks for this. I would like to see something like `localTimeFields` [on the Node](https://github.com/neo4j/graphql/blob/master/packages/graphql/src/classes/Node.ts#L84) and Relationship classes. Then these fields can get picked up in particular places such as [specifying the field has auth on it](https://github.com/neo4j/graphql/blob/f16d2a27b4bc8be088e149658619190379b597d4/packages/graphql/src/classes/Node.ts#L143) and [allowing one to coalesce the field](https://github.com/neo4j/graphql/blob/master/packages/graphql/src/translate/where/create-relationship-where-and-params.ts#L59)(open question?).
> 
> Could they not all be bundled into `dateTimeFields` (maybe rename it `temporalFields`)? I don't think there's really going to be any behaviour differences between the various types.
> 
> We could also get this working with `@timestamp`, which would be cool, but not essential!

I'm not sure if this would work as we do some [specific stuff](https://github.com/neo4j/graphql/blob/master/packages/graphql/src/translate/create-projection-and-params.ts#L435) when projecting `dateTimeFields`.

**Comment by darrellwarde** [comment]
> > > Hey @dmoree, Thanks for this. I would like to see something like `localTimeFields` [on the Node](https://github.com/neo4j/graphql/blob/master/packages/graphql/src/classes/Node.ts#L84) and Relationship classes. Then these fields can get picked up in particular places such as [specifying the field has auth on it](https://github.com/neo4j/graphql/blob/f16d2a27b4bc8be088e149658619190379b597d4/packages/graphql/src/classes/Node.ts#L143) and [allowing one to coalesce the field](https://github.com/neo4j/graphql/blob/master/packages/graphql/src/translate/where/create-relationship-where-and-params.ts#L59)(open question?).
> > 
> > 
> > Could they not all be bundled into `dateTimeFields` (maybe rename it `temporalFields`)? I don't think there's really going to be any behaviour differences between the various types.
> > We could also get this working with `@timestamp`, which would be cool, but not essential!
> 
> I'm not sure if this would work as we do some [specific stuff](https://github.com/neo4j/graphql/blob/master/packages/graphql/src/translate/create-projection-and-params.ts#L435) when projecting `dateTimeFields`.

I actually assumed that `Date` was bundled in there also, but actually, it's "loose" much like `Time` and `LocalTime` are here - we should probably find a way to treat them a bit more uniformly rather than polluting the `Node` and `Relationship` type. There a lot of similarities between all of these types, so maybe a single field on `Node` and `Relationship` with some specific type checks around the codebase is appropriate for any specific behaviours?

**Comment by dmoree** [comment]
@danstarns @darrellwarde I had a similar thought process while creating #449 trying to get `@timestamp` to work with it. I didn't want to change too much and decided to follow a similar pattern with `Date`. If we encompassed them all into `temporalFields`, in order to address @danstarns concern above we could do:
```
...
} else if (temporalField?.typeMeta.name === 'DateTime') {
  res.projection.push(createDatetimeElement({ resolveTree: field, field: temporalField, variable: varName }));
} ...
```

In other words, it is possible to branch off according to `field.typeMeta.name`

Time zones may make it difficult (or impossible) to consistently use `@timestamp` with `Date` or `LocalTime`

**Comment by dmoree** [comment]
> Thanks for also doing LocalTime - nearly there on all of the temporal types now apart from Duration

I could do `Duration` in a similar manner to `Time` and `LocalTime`. I don't work too much with durations, so is there any special considerations, or could I follow the same pattern? My only thought is that perhaps `_GT`, `_LT`, etc. don't really apply here.

**Comment by darrellwarde** [comment]
Hey @dmoree, looks like there's some unresolved references to the removed `validateLocalTime` function, can you take a look? 🙂

**Comment by darrellwarde** [comment]
> @danstarns @darrellwarde I had a similar thought process while creating #449 trying to get `@timestamp` to work with it. I didn't want to change too much and decided to follow a similar pattern with `Date`. If we encompassed them all into `temporalFields`, in order to address @danstarns concern above we could do:
> 
> ```
> ...
> } else if (temporalField?.typeMeta.name === 'DateTime') {
>   res.projection.push(createDatetimeElement({ resolveTree: field, field: temporalField, variable: varName }));
> } ...
> ```
> 
> In other words, it is possible to branch off according to `field.typeMeta.name`
> 
> Time zones may make it difficult (or impossible) to consistently use `@timestamp` with `Date` or `LocalTime`

Exactly, this is how I would go about it - there is enough similarities between them that distinction at this level makes sense.

However, I propose for now that we merge this functionality as it stands (at the end of the day, it's already in-line with the already released `Date` scalar), and then work to bring all 4 temporal types into step as a separate piece of work.

**Comment by darrellwarde** [comment]
> > Thanks for also doing LocalTime - nearly there on all of the temporal types now apart from Duration
> 
> I could do `Duration` in a similar manner to `Time` and `LocalTime`. I don't work too much with durations, so is there any special considerations, or could I follow the same pattern? My only thought is that perhaps `_GT`, `_LT`, etc. don't really apply here.
> 
> Edit: According to the documentation on [Ordering and comparison of values](https://neo4j.com/docs/cypher-manual/current/syntax/operators/#cypher-ordering) `Duration` values cannot be compared and have special considerations for ordering.
> 
> P.S. This thread of the conversation can be moved to another PR.

Definitely a different PR and discussion for `Duration`, and also no rush to get that it, just figured it was worth bringing up! 🙂

**Comment by dmoree** [comment]
> Hey @dmoree, looks like there's some unresolved references to the removed validateLocalTime function

Fixed in https://github.com/neo4j/graphql/pull/451/commits/26567896cd965678e0544a7c84414330149cbfe6.

> Definitely a different PR and discussion for Duration

- #459

**Review by darrellwarde — APPROVED** [review]
LGTM, thanks for all of these contributions! 🙂

**Comment by darrellwarde** [comment]
Necessity for something akin to `temporalFields` captured in #461, merging this as basic support for new temporal types right now.

**Pull request merged**